### ISSUE TITLE
How to define a ReLU with TensorFlow custom_gradient?

### ISSUE THREAD

**Issue by slerman12** [tensorflow/tensorflow#30883]
I'm practicing using TensorFlow's custom_gradient decorator and I tried to define a simple ReLU. One would think it would be as simple as defining the gradient to be 1 when x > 0 and 0 otherwise. However, the following code does not yield the same gradients as a ReLU:

```
@tf.custom_gradient
def relu(x):
    def grad(dy):
        return tf.cond(tf.reshape(x, []) > 0,
                       lambda: tf.cast(tf.reshape(1, dy.shape), tf.float32),
                       lambda: tf.cast(tf.reshape(0, dy.shape), tf.float32))
    return tf.nn.relu(x), grad
```

Can someone explain to me why this standard definition of ReLU's gradient does not yield the same performance as:

```
@tf.custom_gradient
def relu(x):
    def grad(dy):
        return dy
    return tf.nn.relu(x), grad
```

**Comment by gadagashwini** [comment]
@slerman12 Please provide TensorFlow version.Thanks!

**Comment by slerman12** [comment]
1.12.0

**Comment by ymodak** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!

**Comment by slerman12** [comment]
Hi, no one could answer this question on StackOverflow. As far as I know, this is a bug — Why is ReLU’s gradient not 1 when x > 0 and 0 otherwise? That’s the standard definition. Please help, I need to be able to use this custom_gradient function for my research, but it’s not working as expected in this simple example test case.
